File: task_timer.py
import time
import json


class Timer:
    def __init__(self, log_file="timer_logs.json"):
        self.start_time = None
        self.elapsed_time = 0
        self.running = False
        self.log_file = log_file
        self.load_logs()

    def load_logs(self):
        try:
            with open(self.log_file, "r") as file:
                self.logs = json.load(file)
        except (FileNotFoundError, json.JSONDecodeError):
            self.logs = []

    def save_logs(self):
        with open(self.log_file, "w") as file:
            json.dump(self.logs, file, indent=4)

    def start(self):
        if self.running:
            print("The timer is already running.")
        else:
            self.start_time = time.time()
            self.running = True
            print("The timer has started.")

    def stop(self):
        if not self.running:
            print("The timer is not running.")
        else:
            self.elapsed_time += time.time() - self.start_time
            self.start_time = None
            self.running = False
            print(f"The timer has stopped. Your total elapsed time: {self.format_time(self.elapsed_time)}")

    def reset(self):
        if self.running:
            print("Stop the timer before resetting.")
        else:
            self.elapsed_time = 0
            print("Timer reset.")

    def log_session(self, task_name):
        if not self.running:
            print("You cannot log a task while the timer is not running.")
            return

        while not task_name.strip():
            print("Task name cannot be empty. Please try again.")
            task_name = input("Enter task name: ").strip()

        if self.running:
            self.stop()
        session = {
            "Your task": task_name,
            "Time spent on the task": self.format_time(self.elapsed_time),
        }
        self.logs.append(session)
        self.save_logs()
        self.elapsed_time = 0
        print(f"Task '{task_name}' has been logged successfully.")

    def format_time(self, seconds):
        mins, secs = divmod(int(seconds), 60)
        hrs, mins = divmod(mins, 60)
        return f"{hrs:02}:{mins:02}:{secs:02}"

    def show_logs(self):
        if not self.logs:
            print("No logs available.")
        else:
            for i, log in enumerate(self.logs, 1):
                print(f"{i}. Task: {log['Your task']}, Time Spent: {log['Time spent on the task']}")

    def delete_logs(self):
        if not self.logs:
            print("No logs available.")
        else:
            print("Which task would you like to delete?")
            try:
                index = int(input("Enter task number to delete: ").strip())
                self.logs.pop(index - 1)
                self.save_logs()
                print("Log deleted successfully.")
            except (ValueError, IndexError):
                print("Invalid input. Please enter a valid task number.")

    def clear_logs(self):
        if not self.logs:
            print("Nothing to delete.")
        else:
            print("Are you sure you want to delete all logs? (Y/N)")
            choice = input().strip().lower()
            if choice == "y":
                self.logs = []
                self.save_logs()
                print("All logs have been cleared.")
            elif choice == "n":
                print("Logs not cleared.")
            else:
                print("Invalid input.")

    def menu(self):
        print("\nSelect an action: start, stop, reset, log, show, delete, clear, quit")
        while True:
            command = input("Enter command: ").strip().lower()
            if command == "start":
                self.start()
            elif command == "stop":
                self.stop()
            elif command == "reset":
                self.reset()
            elif command == "log":
                task_name = input("Enter task name: ").strip()
                self.log_session(task_name)
            elif command == "show":
                self.show_logs()
            elif command == "delete":
                self.delete_logs()
            elif command == "clear":
                self.clear_logs()
            elif command == "quit":
                if self.elapsed_time > 0 or self.running:
                    if self.running:
                        self.stop()
                    self.log_session("N/A")
                print("Exiting the application.")
                break
            else:
                print("Invalid command. Please try again.")


if __name__ == "__main__":
    app = Timer()
    app.menu()

File: taskTimer.js
const fs = require('fs');
const readlineSync = require('readline-sync');

class TaskTimer {
    constructor(logFile = "timer_logs.json") {
        this.startTime = null;
        this.elapsedTime = 0;
        this.running = false;
        this.logFile = logFile;
        this.loadLogs();
    }

    loadLogs() {
        try {
            if (fs.existsSync(this.logFile)) {
                this.logs = JSON.parse(fs.readFileSync(this.logFile));
            } else {
                this.logs = [];
            }
        } catch (err) {
            this.logs = [];
        }
    }

    saveLogs() {
        fs.writeFileSync(this.logFile, JSON.stringify(this.logs, null, 4));
    }

    start() {
        if (this.running) {
            console.log("The timer is already running.");
        } else {
            this.startTime = Date.now();
            this.running = true;
            console.log("The timer has started.");
        }
    }

    stop() {
        if (!this.running) {
            console.log("The timer is not running.");
        } else {
            this.elapsedTime += (Date.now() - this.startTime) / 1000;
            this.startTime = null;
            this.running = false;
            console.log(`The timer has stopped. Your total elapsed time: ${this.formatTime(this.elapsedTime)}`);
        }
    }

    reset() {
        if (this.running) {
            console.log("Stop the timer before resetting.");
        } else {
            this.elapsedTime = 0;
            console.log("Timer reset.");
        }
    }

    logSession(taskName) {

    
        if (this.elapsedTime === 0) {
            console.log("Cannot save an empty log. No time has been recorded.");
            return;
        }
    
        if (!taskName.trim()) {
            console.log("Task name cannot be empty.");
            return;
        }
    
        if (this.running) this.stop();
    
        const session = {
            "Your task": taskName,
            "Time spent on the task": this.formatTime(this.elapsedTime),
        };
    
        this.logs.push(session);
        this.saveLogs();
        this.elapsedTime = 0;
        console.log(`Task '${taskName}' has been logged successfully.`);
    }
    
    showLogs() {
        if (this.logs.length === 0) {
            console.log("No logs available.");
        } else {
            this.logs.forEach((log, index) => {
                console.log(`${index + 1}. Task: ${log['Your task']}, Time Spent: ${log['Time spent on the task']}`);
            });
        }
    }

    deleteLog() {
        if (this.logs.length === 0) {
            console.log("No logs available.");
            return;
        }
        console.log("Which task would you like to delete? Enter the task number:");
        this.showLogs();
        const index = parseInt(readlineSync.question("Enter task number to delete: ")) - 1;
        if (index >= 0 && index < this.logs.length) {
            this.logs.splice(index, 1);
            this.saveLogs();
            console.log("Log deleted.");
        } else {
            console.log("Invalid task number.");
        }
    }

    clearLogs() {
        if (this.logs.length === 0) {
            console.log("Nothing to delete.");
            return;
        }
        const confirmation = readlineSync.question("Are you sure you want to delete all logs? y/n ");
        if (confirmation.toLowerCase() === 'y'){
            this.logs = [];
            this.saveLogs();
            console.log("All logs cleared.");
        } else {
            console.log("Logs were not cleared.");
        }
    }

    menu() {
        console.log("\nSelect what you want to do: start, stop, reset, log, show, delete, clear, quit");
        const commands = {
            start: () => this.start(),
            stop: () => this.stop(),
            reset: () => this.reset(),
            log: () => {
                const taskName = readlineSync.question("Enter task name: ");
                this.logSession(taskName);
            },
            show: () => this.showLogs(),
            delete: () => this.deleteLog(),
            clear: () => this.clearLogs(),
            quit: () => {
                if (this.running) this.stop();
                console.log("Exiting the application.");
                process.exit(0);
            },
        };

        while (true) {
            const command = readlineSync.question("Enter command: ").toLowerCase();
            if (commands[command]) {
                commands[command]();
            } else {
                console.log("Invalid command. Try again.");
            }
        }
    }

    formatTime(seconds) {
        const hours = Math.floor(seconds / 3600);
        const minutes = Math.floor((seconds % 3600) / 60);
        const remainingSeconds = Math.floor(seconds % 60);
        
        return `${String(hours).padStart(2, '0')}:${String(minutes).padStart(2, '0')}:${String(remainingSeconds).padStart(2, '0')}`;
    }
}

const app = new TaskTimer();
app.menu();
